dlib build failure due to use of _N in opencv headers clashing with _N define from ctype.h

##### System information (version)
- OpenCV => master
- Operating System / Platform => Ubuntu 16.04
- Compiler => Android NDK, standalone arm toolchain api level 21

##### Detailed description

Building dlib with opencv master I just hit errors like:
```
  FAILED: /home/rib/local/android-ndk-r14b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++  --target=armv7-none-linux-androideabi --gcc-toolchain=/home/rib/local/android-ndk-r14b/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64 --sysroot=/home/rib/local/android-ndk-r14b/platforms/android-18/arch-arm  -DOS_ANDROID -Dandroid_dlib_EXPORTS -I../../../../dlib -isystem ../../../../third_party/OpenCV-android-sdk/sdk/native/jni/include -isystem ../../../../third_party/OpenCV-android-sdk/sdk/native/jni/include/opencv -I../../../../third_party/miniglog -I../../../../jni -I../../../../jni/jni_detections -I../../../../include -isystem /home/rib/local/android-ndk-r14b/sources/cxx-stl/llvm-libc++/include -isystem /home/rib/local/android-ndk-r14b/sources/android/support/include -isystem /home/rib/local/android-ndk-r14b/sources/cxx-stl/llvm-libc++abi/include -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -fno-integrated-as -mthumb -Wa,--noexecstack -Wformat -Werror=format-security -std=c++11 -frtti -fexceptions  -std=c++11 -Wall -Werror -O0 -fno-limit-debug-info  -fPIC -MD -MT CMakeFiles/android_dlib.dir/jni/jni_common/jni_bitmap2mat.cpp.o -MF CMakeFiles/android_dlib.dir/jni/jni_common/jni_bitmap2mat.cpp.o.d -o CMakeFiles/android_dlib.dir/jni/jni_common/jni_bitmap2mat.cpp.o -c /mnt/c/Users/Robert/src/glimpse/glimpse-native/dlib/jni/jni_common/jni_bitmap2mat.cpp
  In file included from /mnt/c/Users/Robert/src/glimpse/glimpse-native/dlib/jni/jni_common/jni_bitmap2mat.cpp:10:
  In file included from /mnt/c/Users/Robert/src/glimpse/glimpse-native/dlib/jni/jni_common/jni_bitmap2mat.h:6:
  In file included from ../../../../third_party/OpenCV-android-sdk/sdk/native/jni/include/opencv2/core/core.hpp:48:
  In file included from ../../../../third_party/OpenCV-android-sdk/sdk/native/jni/include/opencv2/core.hpp:59:
  ../../../../third_party/OpenCV-android-sdk/sdk/native/jni/include/opencv2/core/mat.hpp:195:40: error: expected ',' or '>' in template-parameter-list
      template<typename _Tp, std::size_t _N> _InputArray(const std::array<_Tp, _N>& arr);
                                         ^
  /home/rib/local/android-ndk-r14b/platforms/android-18/arch-arm/usr/include/ctype.h:47:12: note: expanded from macro '_N'
  #define _N      0x04
```

It looks like this issue was recently introduced as part of 636ab095b0dda4f63da71e681e47d5e69e6be6b7

--- a/modules/core/include/opencv2/core/mat.hpp
+++ b/modules/core/include/opencv2/core/mat.hpp
@@ -192,8 +192,8 @@ public:
     _InputArray(const std::vector<UMat>& umv);
 
 #ifdef CV_CXX_STD_ARRAY
-    template<typename _Tp, std::size_t _N> _InputArray(const std::array<_Tp, _N>& arr);
-    template<std::size_t _N> _InputArray(const std::array<Mat, _N>& arr);
+    template<typename _Tp, std::size_t _Nm> _InputArray(const std::array<_Tp, _Nm>& arr);
+    template<std::size_t _Nm> _InputArray(const std::array<Mat, _Nm>& arr);
 #endif
 
     Mat getMat(int idx=-1) const;
@@ -324,10 +324,10 @@ public:
     _OutputArray(const std::vector<UMat>& vec);
 
 #ifdef CV_CXX_STD_ARRAY
-    template<typename _Tp, std::size_t _N> _OutputArray(std::array<_Tp, _N>& arr);
-    template<typename _Tp, std::size_t _N> _OutputArray(const std::array<_Tp, _N>& arr);
-    template<std::size_t _N> _OutputArray(std::array<Mat, _N>& arr);
-    template<std::size_t _N> _OutputArray(const std::array<Mat, _N>& arr);
+    template<typename _Tp, std::size_t _Nm> _OutputArray(std::array<_Tp, _Nm>& arr);
+    template<typename _Tp, std::size_t _Nm> _OutputArray(const std::array<_Tp, _Nm>& arr);
+    template<std::size_t _Nm> _OutputArray(std::array<Mat, _Nm>& arr);
+    template<std::size_t _Nm> _OutputArray(const std::array<Mat, _Nm>& arr);
 #endif
 
     bool fixedSize() const;
@@ -390,10 +390,10 @@ public:
     _InputOutputArray(const std::vector<UMat>& vec);
 
 #ifdef CV_CXX_STD_ARRAY
-    template<typename _Tp, std::size_t _N> _InputOutputArray(std::array<_Tp, _N>& arr);
-    template<typename _Tp, std::size_t _N> _InputOutputArray(const std::array<_Tp, _N>& arr);
-    template<std::size_t _N> _InputOutputArray(std::array<Mat, _N>& arr);
-    template<std::size_t _N> _InputOutputArray(const std::array<Mat, _N>& arr);
+    template<typename _Tp, std::size_t _Nm> _InputOutputArray(std::array<_Tp, _Nm>& arr);
+    template<typename _Tp, std::size_t _Nm> _InputOutputArray(const std::array<_Tp, _Nm>& arr);
+    template<std::size_t _Nm> _InputOutputArray(std::array<Mat, _Nm>& arr);
+    template<std::size_t _Nm> _InputOutputArray(const std::array<Mat, _Nm>& arr);
 #endif
 
 };
@@ -980,7 +980,7 @@ public:
 #ifdef CV_CXX_STD_ARRAY
     /** @overload
     */
-    template<typename _Tp, size_t _N> explicit Mat(const std::array<_Tp, _N>& arr, bool copyData=false);
+    template<typename _Tp, size_t _Nm> explicit Mat(const std::array<_Tp, _Nm>& arr, bool copyData=false);
 #endif
 
     /** @overload
@@ -1607,7 +1607,7 @@ public:
     template<typename _Tp, int m, int n> operator Matx<_Tp, m, n>() const;
 
 #ifdef CV_CXX_STD_ARRAY
-    template<typename _Tp, std::size_t _N> operator std::array<_Tp, _N>() const;
+    template<typename _Tp, std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
 #endif
 
     /** @brief Reports whether the matrix is continuous or not.
@@ -2156,7 +2156,7 @@ public:
     explicit Mat_(const MatCommaInitializer_<_Tp>& commaInitializer);
 
 #ifdef CV_CXX_STD_ARRAY
-    template <std::size_t _N> explicit Mat_(const std::array<_Tp, _N>& arr, bool copyData=false);
+    template <std::size_t _Nm> explicit Mat_(const std::array<_Tp, _Nm>& arr, bool copyData=false);
 #endif
 
     Mat_& operator = (const Mat& m);
@@ -2255,7 +2255,7 @@ public:
 
 #ifdef CV_CXX_STD_ARRAY
     //! conversion to array.
-    template<std::size_t _N> operator std::array<_Tp, _N>() const;
+    template<std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
 #endif
 
     //! conversion to Vec

--- a/modules/core/include/opencv2/core/mat.inl.hpp
+++ b/modules/core/include/opencv2/core/mat.inl.hpp
@@ -78,13 +78,13 @@ _InputArray::_InputArray(const std::vector<_Tp>& vec)
 { init(FIXED_TYPE + STD_VECTOR + DataType<_Tp>::type + ACCESS_READ, &vec); }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-_InputArray::_InputArray(const std::array<_Tp, _N>& arr)
-{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_READ, arr.data(), Size(1, _N)); }
+template<typename _Tp, std::size_t _Nm> inline
+_InputArray::_InputArray(const std::array<_Tp, _Nm>& arr)
+{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_READ, arr.data(), Size(1, _Nm)); }
 
-template<std::size_t _N> inline
-_InputArray::_InputArray(const std::array<Mat, _N>& arr)
-{ init(STD_ARRAY_MAT + ACCESS_READ, arr.data(), Size(1, _N)); }
+template<std::size_t _Nm> inline
+_InputArray::_InputArray(const std::array<Mat, _Nm>& arr)
+{ init(STD_ARRAY_MAT + ACCESS_READ, arr.data(), Size(1, _Nm)); }
 #endif
 
 inline
@@ -162,13 +162,13 @@ _OutputArray::_OutputArray(std::vector<_Tp>& vec)
 { init(FIXED_TYPE + STD_VECTOR + DataType<_Tp>::type + ACCESS_WRITE, &vec); }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-_OutputArray::_OutputArray(std::array<_Tp, _N>& arr)
-{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_WRITE, arr.data(), Size(1, _N)); }
+template<typename _Tp, std::size_t _Nm> inline
+_OutputArray::_OutputArray(std::array<_Tp, _Nm>& arr)
+{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_WRITE, arr.data(), Size(1, _Nm)); }
 
-template<std::size_t _N> inline
-_OutputArray::_OutputArray(std::array<Mat, _N>& arr)
-{ init(STD_ARRAY_MAT + ACCESS_WRITE, arr.data(), Size(1, _N)); }
+template<std::size_t _Nm> inline
+_OutputArray::_OutputArray(std::array<Mat, _Nm>& arr)
+{ init(STD_ARRAY_MAT + ACCESS_WRITE, arr.data(), Size(1, _Nm)); }
 #endif
 
 inline
@@ -200,13 +200,13 @@ _OutputArray::_OutputArray(const std::vector<_Tp>& vec)
 { init(FIXED_TYPE + FIXED_SIZE + STD_VECTOR + DataType<_Tp>::type + ACCESS_WRITE, &vec); }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-_OutputArray::_OutputArray(const std::array<_Tp, _N>& arr)
-{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_WRITE, arr.data(), Size(1, _N)); }
+template<typename _Tp, std::size_t _Nm> inline
+_OutputArray::_OutputArray(const std::array<_Tp, _Nm>& arr)
+{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_WRITE, arr.data(), Size(1, _Nm)); }
 
-template<std::size_t _N> inline
-_OutputArray::_OutputArray(const std::array<Mat, _N>& arr)
-{ init(FIXED_SIZE + STD_ARRAY_MAT + ACCESS_WRITE, arr.data(), Size(1, _N)); }
+template<std::size_t _Nm> inline
+_OutputArray::_OutputArray(const std::array<Mat, _Nm>& arr)
+{ init(FIXED_SIZE + STD_ARRAY_MAT + ACCESS_WRITE, arr.data(), Size(1, _Nm)); }
 #endif
 
 template<typename _Tp> inline
@@ -277,13 +277,13 @@ _InputOutputArray::_InputOutputArray(std::vector<_Tp>& vec)
 { init(FIXED_TYPE + STD_VECTOR + DataType<_Tp>::type + ACCESS_RW, &vec); }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-_InputOutputArray::_InputOutputArray(std::array<_Tp, _N>& arr)
-{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_RW, arr.data(), Size(1, _N)); }
+template<typename _Tp, std::size_t _Nm> inline
+_InputOutputArray::_InputOutputArray(std::array<_Tp, _Nm>& arr)
+{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_RW, arr.data(), Size(1, _Nm)); }
 
-template<std::size_t _N> inline
-_InputOutputArray::_InputOutputArray(std::array<Mat, _N>& arr)
-{ init(STD_ARRAY_MAT + ACCESS_RW, arr.data(), Size(1, _N)); }
+template<std::size_t _Nm> inline
+_InputOutputArray::_InputOutputArray(std::array<Mat, _Nm>& arr)
+{ init(STD_ARRAY_MAT + ACCESS_RW, arr.data(), Size(1, _Nm)); }
 #endif
 
 inline _InputOutputArray::_InputOutputArray(std::vector<bool>&)
@@ -314,13 +314,13 @@ _InputOutputArray::_InputOutputArray(const std::vector<_Tp>& vec)
 { init(FIXED_TYPE + FIXED_SIZE + STD_VECTOR + DataType<_Tp>::type + ACCESS_RW, &vec); }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-_InputOutputArray::_InputOutputArray(const std::array<_Tp, _N>& arr)
-{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_RW, arr.data(), Size(1, _N)); }
+template<typename _Tp, std::size_t _Nm> inline
+_InputOutputArray::_InputOutputArray(const std::array<_Tp, _Nm>& arr)
+{ init(FIXED_TYPE + FIXED_SIZE + STD_ARRAY + DataType<_Tp>::type + ACCESS_RW, arr.data(), Size(1, _Nm)); }
 
-template<std::size_t _N> inline
-_InputOutputArray::_InputOutputArray(const std::array<Mat, _N>& arr)
-{ init(FIXED_SIZE + STD_ARRAY_MAT + ACCESS_RW, arr.data(), Size(1, _N)); }
+template<std::size_t _Nm> inline
+_InputOutputArray::_InputOutputArray(const std::array<Mat, _Nm>& arr)
+{ init(FIXED_SIZE + STD_ARRAY_MAT + ACCESS_RW, arr.data(), Size(1, _Nm)); }
 #endif
 
 template<typename _Tp> inline
@@ -558,8 +558,8 @@ Mat::Mat(const std::vector<_Tp>& vec, bool copyData)
 }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-Mat::Mat(const std::array<_Tp, _N>& arr, bool copyData)
+template<typename _Tp, std::size_t _Nm> inline
+Mat::Mat(const std::array<_Tp, _Nm>& arr, bool copyData)
     : flags(MAGIC_VAL | DataType<_Tp>::type | CV_MAT_CONT_FLAG), dims(2), rows((int)arr.size()),
       cols(1), data(0), datastart(0), dataend(0), allocator(0), u(0), size(&rows)
 {
@@ -1197,10 +1197,10 @@ Mat::operator std::vector<_Tp>() const
 }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp, std::size_t _N> inline
-Mat::operator std::array<_Tp, _N>() const
+template<typename _Tp, std::size_t _Nm> inline
+Mat::operator std::array<_Tp, _Nm>() const
 {
-    std::array<_Tp, _N> v;
+    std::array<_Tp, _Nm> v;
     copyTo(v);
     return v;
 }
@@ -1561,8 +1561,8 @@ Mat_<_Tp>::Mat_(const std::vector<_Tp>& vec, bool copyData)
 {}
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp> template<std::size_t _N> inline
-Mat_<_Tp>::Mat_(const std::array<_Tp, _N>& arr, bool copyData)
+template<typename _Tp> template<std::size_t _Nm> inline
+Mat_<_Tp>::Mat_(const std::array<_Tp, _Nm>& arr, bool copyData)
     : Mat(arr, copyData)
 {}
 #endif
@@ -1845,10 +1845,10 @@ Mat_<_Tp>::operator std::vector<_Tp>() const
 }
 
 #ifdef CV_CXX_STD_ARRAY
-template<typename _Tp> template<std::size_t _N> inline
-Mat_<_Tp>::operator std::array<_Tp, _N>() const
+template<typename _Tp> template<std::size_t _Nm> inline
+Mat_<_Tp>::operator std::array<_Tp, _Nm>() const
 {
-    std::array<_Tp, _N> a;
+    std::array<_Tp, _Nm> a;
     copyTo(a);
     return a;
 }

